app/controllers/users/omniauth_callbacks_controller.rb

```diff
@@ -9,71 +12,109 @@ class OmniauthCallbacksController < Devise::OmniauthCallbacksController
     # SAML doesn't support the CSRF protection
     protect_from_forgery except: %i[mocksaml bird nbp]
 
-    def sso_callback # rubocop:disable Metrics/AbcSize
-      # Check if an existing user is already signed in (passed through the RelayState)
-      # and trying to add a new identity to their account. If so, we load the user information
-      # and set it as the current user. This is necessary to avoid creating a new user.
-      current_user = User.find_by(id: OmniAuth::NonceStore.pop(params[:RelayState]))
-
-      # For existing users, we want to redirect to their profile page after adding a new identity
-      store_location_for(:user, edit_user_registration_path) if current_user.present?
-
-      # The instance variable `@user` is used by Devise internally and should be set here
-      @user = User.from_omniauth(request.env['omniauth.auth'], current_user)
-
-      if @user.persisted?
-        # The `sign_in_and_redirect` will only proceed with the login if the account has been confirmed
-        # either through trust by the IdP or with a confirmation mail. Until then, a flash message
-        # notifies users about the missing email confirmation
-        sign_in_and_redirect @user, event: :authentication
-        # We use the configured OmniAuth camilization to include the user-facing name of the provider in the flash message
-        set_flash_message(:notice, :success, kind: OmniAuth::Utils.camelize(provider)) if is_navigational_format?
-        session['omniauth_provider'] = provider
+    def sso_callback
```

Comment:
When performing a regular sign-in, we do not store the `session[:omniauth_provider]` in the current version. This is problematic, since it breaks the Single Log-Out workflow. Thus, my recommendation would be to store the value first. Then, we can rely on all three information (`saml_uid`, `saml_session_index`, `omniauth_provider`) to be present.

You might argue that in case of an error we would than not remove the `omniauth_provider` (since it is simply not set in the current implementation). That's true, but also an issue for the other two and not yet tackled for these either.